waspc/test/AnalyzerTest.hs

```diff
@@ -189,6 +191,74 @@ spec_Analyzer = do
               ]
       isRight (analyze source) `shouldBe` True
 
+    describe "Returns correct error message" $ do
+      let errorMessageShouldBe analyzeResult (pos, msg) = case analyzeResult of
+            Right _ -> error "Test failed: expected AnalyzerError."
+            Left e -> getErrorMessageAndPosition e `shouldBe` (msg, pos)
+
+      it "For nested unexpected type error" $ do
+        let source =
+              unlines
+                [ "app MyApp {",
+                  "  title: \"My app\",",
+                  "  dependencies: [",
+                  "    { name: \"bar\", version: 13 },",
+                  "    { name: \"foo\", version: 14 }",
+                  "  ]",
+                  "}"
+                ]
+        analyze source
+          `errorMessageShouldBe` ( SourcePosition 4 29,
+                                   intercalate
+                                     "\n"
+                                     [ "Type error:",
+                                       "  For dictionary field 'dependencies':",
+                                       "    For list element:",
+                                       "      For dictionary field 'version':",
```

Comment:
I could remove this additional information -> I am not sure if it is useful or not. We can skip printing it we wish. WDYT?

Reply:
I think for this first go around it is fine as is. I'm trying to think of any way to more concisely say it that also scales to arbitrary structures and it is hard to do so :) I think having the surrounding context makes it a bit easier to visualize where/what the problem is.

Reply:
I was thinking the same thing! It will be easier to judge once we actually start using this new Analyzer in Wasp, so we can see then what to do.

Reply:
I think it might be helpful to turn it upside down - first we start with the bottom leaf (e.g. "type error for dict field version") and then in the lines below continue with location details ("in list -> in field 'dependencies' -> ...). That way dev can immediately identify the place of error but then also get more details, other than starting from the top leaf and digging down until he gets to the actual thing.

I think I have also seen error messages structured like that in other languages, even in Haskell if I am not wrong? @shayneczyzewski @Martinsos what is your experience?

Reply:
You are right @matijaSos !

I imagined that they would read it bottom - up -> I personally usually look at the very end of the error message first and then read from there up. But I wonder now if that is the best way hm.

In Haskell, they do as you suggest. First comes the error message, then comes this "in", "for", but from inside out, they way you suggested it.
So, following that, we should print smth like:
```hs
"Unexpected type 'foo'
" └> For dictionary field 'version':",
"   └> In list:",
"     └> For dictionary field 'dependencies':",
```

Maybe we should research couple other languages also, to make sure what is the usual way, and also research what people like the most. But this does make sense.

Is it ok if I create separate issue for this? This is a nice, cool standalone problem that could be a great entry point for somebody to start playing with compiler!

Reply:
Created issue for it here: https://github.com/wasp-lang/wasp/issues/414 .